proxy/tcpproxy/userspace.go

```diff
@@ -90,6 +97,61 @@ func (tp *TCPProxy) Run() error {
 	}
 }
 
+type PickerFunc func(tp *TCPProxy) *remote
+
+// NewPicker returns a picker to choose remotes for the tcp proxy.
+func NewPicker(name string) (PickerFunc, error) {
+	switch name {
+	case "srv-priority":
+		return srvPicker, nil
+	case "round-robin":
+		nextRemote := 0
+		f := func(tp *TCPProxy) *remote {
+			for i := 0; i < len(tp.remotes); i++ {
+				picked := tp.remotes[nextRemote]
+				nextRemote = (nextRemote + 1) % len(tp.remotes)
+				if picked.isActive() {
+					return picked
+				}
+			}
+			return nil
+		}
+		return f, nil
+	default:
+	}
+	return nil, fmt.Errorf("unknown picker %q", name)
+}
+
+func srvPicker(tp *TCPProxy) *remote {
+	bestPr := uint16(65535)
+	w := 0
+	var candidates []*remote
+	// find best priority class
+	for _, r := range tp.remotes {
+		switch {
+		case !r.isActive():
+		case r.srv.Priority < bestPr:
+			bestPr = r.srv.Priority
+			candidates = []*remote{r}
+			w = int(r.srv.Weight) + 1
+		case r.srv.Priority == bestPr:
+			candidates = append(candidates, r)
+			w += int(r.srv.Weight) + 1
+		}
+	}
+	// randomly choose by weight
```

Comment:
>  A server selection mechanism.  The weight field specifies a
        relative weight for entries with the same priority. Larger
        weights SHOULD be given a proportionately higher probability of
        being selected. The range of this number is 0-65535.  This is a
        16 bit unsigned integer in network byte order.  Domain
        administrators SHOULD use Weight 0 when there isn't any server
        selection to do, to make the RR easier to read for humans (less
        noisy).  In the presence of records containing weights greater
        than 0, records with weight 0 should have a very small chance of
        being selected.

 Can we follow this?